ec_deployment : Tags for deployment datasource

neiljbrookes opened the pull request:
<!--- Provide a general summary of your changes in the title above. -->

## Description
<!--- Describe your changes in detail. -->

This PR surafces any metadata tags associated with a deployment datasource.

- Update `schema.go` to include "tags" in the ec_deployment schema
- Update `datasrouce.go` (modelToState) to operate on any "tags" present in the model.
- Add `flatteners_tags.go` (flattenTags) a function to flatten the model tags into a flat map.
- Add `flatteners_tags_test.go` A unit test.
- Update `datasource_test.go`. Adddition of tag flattening to the test.
- Update `ec_deployment.md`. Includes "tags" as an attribute reference in the docs.


## Related Issues
<!--- This project only accepts pull requests related to open issues. -->
<!--- If suggesting a new feature or change, please discuss it in an -->
<!--- issue first.  If fixing a bug, there should be an issue describing -->
<!--- it with steps to reproduce.  Please link to the any related issues -->
<!--- here: -->

Closes : #182

## Motivation and Context
<!--- Why is this change required? What problem does it solve? -->

- Metadata tags as first class citizens.
- Parity between resource and data_source.

## How Has This Been Tested?
<!--- Please describe in detail how you tested your changes.  Include -->
<!--- details of your testing environment, and the tests you ran to -->
<!--- see how your change affects other areas of the code, etc. -->

Unit and manual tests.

I have used this defintion to test.

<details>

```tf
resource "ec_deployment" "resource" {
  name = "resource"

  region                 = "aws-eu-west-1"
  version                = "7.10.1"
  deployment_template_id = "aws-cross-cluster-search-v2"

  elasticsearch {}
  kibana {}
  tags = {
    "foo"   = "bar",
    "bar"   = "baz"
  }
}

data "ec_deployment" "datasource" {
  id = ec_deployment.resource.id
}

output "debug" {
  description = "test"
  value = data.ec_deployment.datasource.tags
}
```
</details>

## Types of Changes
<!--- What types of changes does your code introduce? Put an `x` in -->
<!--- all the boxes that apply: -->
- [ ] Bug fix (non-breaking change which fixes an issue)
- [x] New feature (non-breaking change which adds functionality)
- [ ] Refactoring (improves code quality but has no user-facing effect)
- [ ] Breaking change (fix or feature that would cause existing functionality to change)
- [ ] Documentation

## Readiness Checklist
<!--- Go over all the following points, and put an `x` in all the boxes -->
<!--- that apply.  If you're unsure about any of these, don't hesitate -->
<!--- to ask. We're here to help! -->
- [x] My code follows the code style of this project
- [x] My change requires a change to the documentation
- [x] I have updated the documentation accordingly
- [x] I have added tests to cover my changes
- [x] All new and existing tests passed

karencfv commented on docs/data-sources/ec_deployment.md:
```diff
@@ -32,6 +32,7 @@ These will not be available for interpolation.
 * `deployment_template_id` - ID of the deployment template used to create the deployment.
 * `traffic_filter` - Traffic filter block, which contains a list of traffic filter rule identifiers.
 * `observability` Observability settings. Information about logs and metrics shipped to a dedicated deployment.
+* `tags` Metadata tags. Arbitrary tags applied to the deployment.
```
Let's move this line above the `observability` line because the lines below are part of the observability object. Also we should probably change the text to say what it says on the resource [description](https://github.com/elastic/terraform-provider-ec/blob/master/docs/resources/ec_deployment.md#argument-reference).
```suggestion
* `tags` Key value map of arbitrary string tags..
```

karencfv reviewed (commented):
Thanks @neiljbrookes for opening up the PR! 🎉 🙇‍♀️ 

The code looks great! Apart from the docs comment, we should add this feature to the acceptance tests.

To modify the deployment datasource acc test you need to do two things:
1. Add a tag to the test .tf file [here](https://github.com/elastic/terraform-provider-ec/blob/master/ec/acc/testdata/datasource_deployment_basic.tf#L42)
2. Verify that the tag on the datasource is the same as the one in the resource [here](https://github.com/elastic/terraform-provider-ec/blob/master/ec/acc/datasource_deployment_basic_test.go#L51)

The resource tags [test](https://github.com/elastic/terraform-provider-ec/blob/master/ec/acc/deployment_basic_tags_test.go) can help as a guideline.

You can take a look at the resource [tags test](https://github.com/elastic/terraform-provider-ec/blob/master/ec/acc/deployment_basic_tags_test.go). It will be similar.

To run this test locally you would need to run `make testacc TEST_NAME=TestAccDatasourceDeployment_basic`.

**Warning** While this test is passing on master it's going through a "getting it fixed" state , this [PR](https://github.com/elastic/terraform-provider-ec/pull/242) is the second part of getting it fixed. So you might want to wait for it to get merged so you can pull the changes otherwise you might also get some annoying merge conflicts 😄

neiljbrookes commented:
Rebased from master to get https://github.com/elastic/terraform-provider-ec/pull/242. 
Force pushed.

karencfv commented on docs/data-sources/ec_deployment.md:
```diff
@@ -32,6 +32,7 @@ These will not be available for interpolation.
 * `deployment_template_id` - ID of the deployment template used to create the deployment.
 * `traffic_filter` - Traffic filter block, which contains a list of traffic filter rule identifiers.
 * `observability` Observability settings. Information about logs and metrics shipped to a dedicated deployment.
+* `tags` Key value map of arbitrary string tags..
```
I think you forgot to move this line above the observability one 😄

neiljbrookes replied:
Indeed. I just mashed the "commit suggestion" button.
Fixed in https://github.com/elastic/terraform-provider-ec/pull/244/commits/ea5f851dc733a60302b5beae986c59f97b724e6a

alaudazzi reviewed (approved):
LGTM for the doc change on ec_deployment.md.

neiljbrookes commented:
@karencfv I've made an attempt at some acceptance tests !

- Updated `TestAccDatasourceDeployment_basic` to include tags.
- Added `TestAccDatasource_basic_tags` and the associted TF defintion `datasource_tags.tf`.

Its quite simple, like me 😆. 
Just a small "ec_deployment" resource with specific tags is created, then an "ec_deployment" datasource reads that data and the test ensure that the tags are as expected.

Tested OK for me 
```
> make testacc TEST_NAME=TestAccDatasource_basic_tags
-> Running terraform acceptance tests...
=== RUN   TestAccDatasource_basic_tags
=== PAUSE TestAccDatasource_basic_tags
=== CONT  TestAccDatasource_basic_tags
--- PASS: TestAccDatasource_basic_tags (184.86s)
PASS
ok      github.com/elastic/terraform-provider-ec/ec/acc 185.908s
```

The `TestAccDatasourceDeployment_basic` is still failing for me (usually with something that looks like https://github.com/elastic/cloud/issues/75870). Seems to still be affecting master, so I dont think it is somerthing that I broke !

karencfv commented:
Thanks a bunch for addressing my comments @neiljbrookes 🙇‍♀️ 

I tried running the `TestAccDatasource_basic` test with your changes locally several times and unfortunately I kept running into `"plan failed due to unknown error"` error messages from the API which are not related to your PR.

In the interest of moving your PR along I think the `TestAccDatasourceDeployment_basic_tags` test is enough, and we can get rid of changes to `TestAccDatasourceDeployment_basic` and `datasource_deployment_basic.tf`. Those can be added later when we can run the test 😄

neiljbrookes commented:
Thanks @karencfv .
I have removed the "tags" references from `TestAccDatasourceDeployment_basic` and `datasource_deployment_basic.tf`.

To clarify, I get successful tests on `TestAccDatasource_basic_tags` 
```
> make testacc TEST_NAME=TestAccDatasource_basic_tags          
-> Running terraform acceptance tests...
=== RUN   TestAccDatasource_basic_tags
=== PAUSE TestAccDatasource_basic_tags
=== CONT  TestAccDatasource_basic_tags
--- PASS: TestAccDatasource_basic_tags (303.12s)
PASS
ok      github.com/elastic/terraform-provider-ec/ec/acc 304.208s
```

and faillure on `TestAccDatasourceDeployment_basic`
```
> make testacc TEST_NAME=TestAccDatasourceDeployment_basic
-> Running terraform acceptance tests...
=== RUN   TestAccDatasourceDeployment_basic
=== PAUSE TestAccDatasourceDeployment_basic
=== CONT  TestAccDatasourceDeployment_basic
    testing_new.go:63: Error running post-test destroy, there may be dangling resources: exit status 1
        
        Error: found deployment plan errors: 1 error occurred:
                * deployment [3171b9d74a5e4ee379440597c92be238] - [elasticsearch][43108bb53f2548f98c1b25b64eb538c8]: caught error: "plan failed due to unknown error"
        
        
        
        
--- FAIL: TestAccDatasourceDeployment_basic (475.01s)
FAIL
FAIL    github.com/elastic/terraform-provider-ec/ec/acc 476.064s
FAIL
make: *** [testacc] Error 1
```

karencfv reviewed (approved):
Thanks for adding this feature @neiljbrookes 🎉  Everything looks great now

I have confirmed that the errors on the test are unrelated to this PR. We've had some ongoing issues with this suite particularly :(

neiljbrookes commented:
Thanks for all your help @karencfv.
My next challange is to add tag filtering for the `ec_deployments` datasource. I'm sure I'll be annoying you again shortly.
